feat: Show compatibility errors in Disco Pane (fix #2095)

tofumatt opened the pull request:
Working on showing errors in the disco pane; tests are failing right now and it's low-priority so just leaving here now but will revisit later.

muffinresearch commented:
@tofumatt - this looks to have bit-rotted, can it be updated and completed?

muffinresearch commented:
![sad_keanu](https://user-images.githubusercontent.com/1514/27788687-7bba2822-5fe1-11e7-8992-8870a0db10e0.jpg)

tofumatt commented:
Alright, updated and fixed everything, here are some screenshots:

### Small screen
![screenshot 2017-07-08 02 45 11](https://user-images.githubusercontent.com/90871/27986001-cec5e332-63ef-11e7-8771-0156465db44a.png)

### Desktop
![screenshot 2017-07-08 02 44 46](https://user-images.githubusercontent.com/90871/27986002-cec74272-63ef-11e7-8404-dbce8509c19b.png)
![screenshot 2017-07-08 02 44 48](https://user-images.githubusercontent.com/90871/27986000-cec47fd8-63ef-11e7-8d4c-c5255b3975de.png)

muffinresearch commented on src/disco/components/AddonCompatibilityError/style.scss:
```diff
@@ -0,0 +1,16 @@
+
+  &,
+  a,
+  a:link {
```
Any other states needed e.g: visited, hover, focus, active  etc or are the defaults ok?

tofumatt replied:
The defaults are fine, this just keeps the colour white which it will stay for those other states too.

muffinresearch commented on src/amo/components/AddonReviewList.js:
```diff
@@ -120,6 +120,7 @@ export class AddonReviewListBase extends React.Component {
           { authorName: review.userName, timestamp });
 
       const reviewBodySanitized = sanitizeHTML(nl2br(review.body), ['br']);
```
Is this intermediate `const` needed?

tofumatt replied:
It tidies up the next lines and makes it a bit more readable. I went back and forth on it and this was the nicest version.

muffinresearch commented on src/disco/components/AddonCompatibilityError/index.js:
```diff
@@ -0,0 +1,87 @@
+      throw new Error('minVersion is required; it cannot be undefined');
+    }
+
+    if (reason === INCOMPATIBLE_NOT_FIREFOX) {
```
Should never really happen since non non-FF browsers should ever see the disco pane.

tofumatt replied:
True, but they could hit the URL and it doesn't hurt to have nice messages. Could also happen if the UA was weird and might be useful in a bug report...

That said it increases code size. If you want, I can remove this if we care a lot less about other clients in the Disco Pane, which I guess we do.

muffinresearch commented on tests/unit/disco/components/TestAddon.js:
```diff
@@ -319,8 +359,9 @@ describe('<Addon />', () => {
+
       expect(preventDefault.called).toBeTruthy();
-      expect(installTheme.calledWith(themeImage, data.addon)).toBeTruthy();
+      sinon.assert.calledWith(installTheme, themeImage);
```
is `data.addon` no longer passed?

tofumatt replied:
This was actually a case of "how did this test ever work?" since updating the `withInstallButton` HOC. I fixed it up using enzyme... but we should definitely clean up this test suite in general...

muffinresearch commented:
This is probably a @pwalm question but I wonder if the red could be dropped back a bit? The sea of red for a non-FF browser might be a bit much.

muffinresearch commented on tests/unit/disco/components/TestAddon.js:
```diff
@@ -143,22 +141,31 @@ describe('<Addon />', () => {
       const root = renderAddon({ addon: result, ...result });
-      expect(root.heading.textContent).toContain('test-heading');
+
+      expect(findDOMNode(root).querySelector('.heading').textContent)
```
Iirc `findDOMNode` was planned to be deprecated see https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-find-dom-node.md

See also https://facebook.github.io/react/docs/react-dom.html#finddomnode re: use being discouraged.

tofumatt replied:
It is, and we'll be moving to enzyme, but I didn't want to refactor these tests at the time. I could do it now if you like but this isn't our only usage of `findDomNode` 😄

muffinresearch commented on tests/unit/disco/components/TestAddonCompatibilityError.js:
```diff
@@ -0,0 +1,92 @@
+
+  it('renders a notice for old versions of Firefox', () => {
+    const root = render({
+      lang: 'fr',
```
Seems a little odd to pass lang and then be looking for english translations.

tofumatt replied:
Oh that's true, I'll tweak that.

pwalm commented:
The red is a bit much, isn't it. I would suggest we put that "You need to..." notification between the intro copy and the first add-on, and remove or make inactive the "Add to Firefox"/"Install Theme" buttons.
